Set permissions for flatpaks

meisenzahl opened the pull request:
# Description
As described in the [AppCenter for Everyone](https://www.indiegogo.com/projects/appcenter-for-everyone#/) campaign, you want to use Flatpak as a format for apps in the future. So I thought it would be a good idea to be able to set the permissions of the apps. By the way: Congratulations on the success of the campaign 🎉

For the implementation I have oriented myself on [Flatseal](https://github.com/tchx84/Flatseal/). Great foundation!

The extension of the plug now offers the same functionality as [Flatseal](https://github.com/tchx84/Flatseal/), except for the setting of customized paths. But I am already working on that ;)

I created the pull request to get feedback from you.

# Current state

![Screenshot von 2020-02-21 09 40 28](https://user-images.githubusercontent.com/10796736/75018886-31750d80-5490-11ea-9ad7-eb47d4e68962.png)

![Screenshot von 2020-02-21 09 42 21](https://user-images.githubusercontent.com/10796736/75018903-3afe7580-5490-11ea-9420-0fb01f753f2d.png)

The permissions are immediately overwritten when toggling a switch. For example, [Games](https://flathub.org/apps/details/org.gnome.Games) can't start anymore if I revoke the permission to access X11:

```shell
Gtk-Message: 09:42:00.497: Failed to load module "pantheon-filechooser-module"
Unable to init server: Could not connect: Connection refused

(org.gnome.Games:2): Gtk-WARNING **: 09:42:00.506: cannot open display:
```

![Screenshot von 2020-02-21 09 42 28](https://user-images.githubusercontent.com/10796736/75018920-418ced00-5490-11ea-8f63-1c45c5e3574b.png)

Clicking the **Reset** button restores the default permissions.

# Tasks

- [ ] Redesign the GUI
  - [ ] Minimize the **Reset** button
  - [ ] Minimize the size of the switches
- [ ] Refactor the code

# Questions

- I created an implementation of a **Map** because [GLib.HashTable](https://valadoc.org/glib-2.0/GLib.HashTable.html) changes the order of the keys. Would it be okay to add [Gee](https://valadoc.org/gee-0.8/Gee.html) as a dependency?

I am looking forward to your feedback!

davidmhewitt commented:
Wow, this is a really interesting starting point! Thank you for this.

It would be fine to add Gee as a dependency, we use this nearly everywhere so one more place won't hurt :smile: 

One of the things that we're going to have to think about at the sprint is a way to write the description of all of these permissions so that they're clear to the average user. For example, I shouldn't have to know what "inter-process communications" means. I should just be told it means that apps may be able to talk to, control and read information from other apps or system components.

Another thing worth considering for us is that we're going to need to at least display these permissions in at least 3 applications (Sideload, AppCenter and here in the applications plug), so we may consider building a small library for reading, translating/describing, and setting these permissions rather than having this code duplicated in 3+ places.

I think it's unlikely we'll get around to reviewing this properly or merging it until during/after the sprint. The first priority is building the flatpak authenticator, the backend infrastructure for building flatpaks into the new elementary repo and maybe some flatpak portals. But, thank you for all your hard work! It'll certainly save us some research :heart:

meisenzahl commented:
Many thanks for the detailed feedback!

To make the effects of permissions clearer to users, they may need to be grouped by topic. However, this adds a whole new level of complexity because, for example, graphical output is allowed for an app using X11, but not using Wayland. Mapping this to the flatpak permissions is a challenge.

I think developing a library for handling the Flatpak permissions is a good idea. The logic for handling the permissions is already implemented. Developing a reusable GUI for it is still pending. I trust in the capabilities of the elementary project 😉️

If I can help in any way, please let me know.

meisenzahl commented:
> It would be fine to add Gee as a dependency, we use this nearly everywhere so one more place won't hurt

Apparently there is no `Map` in [Gee](https://valadoc.org/gee-0.8/Gee.html) that keeps the order of the keys. So I leave my implementation with 2 internal lists like this.

vjr commented:
> > It would be fine to add Gee as a dependency, we use this nearly everywhere so one more place won't hurt
> 
> Apparently there is no `Map` in [Gee](https://valadoc.org/gee-0.8/Gee.html) that keeps the order of the keys. So I leave my implementation with 2 internal lists like this.

Won't the TreeMap work here? See https://valadoc.org/gee-0.8/Gee.TreeMap.html

vjr commented:
> > > It would be fine to add Gee as a dependency, we use this nearly everywhere so one more place won't hurt
> > 
> > 
> > Apparently there is no `Map` in [Gee](https://valadoc.org/gee-0.8/Gee.html) that keeps the order of the keys. So I leave my implementation with 2 internal lists like this.
> 
> Won't the TreeMap work here? See https://valadoc.org/gee-0.8/Gee.TreeMap.html

My bad, probably not.

danrabbit commented on src/Permissions/Widgets/PermissionSettingsWidget.vala:
```diff
@@ -0,0 +1,57 @@
+
+        pack_start (vbox, false, false, 0);
+
+        var s = new Gtk.Switch ();
```
The fix for the switch size is to set `valign = Gtk.Align.Center` :)

danrabbit commented on src/Permissions/Widgets/PermissionSettingsWidget.vala:
```diff
@@ -0,0 +1,57 @@
+
+        orientation = Gtk.Orientation.HORIZONTAL;
+        hexpand = true;
+        var vbox = new Gtk.Box (Gtk.Orientation.VERTICAL, 0);
```
Instead of nesting boxes, it's usually better to use Gtk.Grid: https://valadoc.org/gtk+-3.0/Gtk.Grid.html

meisenzahl commented:
@danrabbit many thanks for your detailed feedback and constructive suggestions! 

As suggested, I have removed the options for X11 fallback and Wayland, as they are not needed.

I implemented your UI suggestions and added icons and descriptions to the permissions view. I find the UI now more appealing. But I think you probably have more good ideas to make it even more user friendly 😉️

![Screenshot von 2020-02-22 09 19 06](https://user-images.githubusercontent.com/10796736/75089009-a90c7080-5554-11ea-8ea8-615e7b4cacca.png)

![Screenshot von 2020-02-22 09 20 02](https://user-images.githubusercontent.com/10796736/75089032-e1ac4a00-5554-11ea-81b3-0861f4558779.png)

![Screenshot von 2020-02-22 09 20 06](https://user-images.githubusercontent.com/10796736/75089033-e40ea400-5554-11ea-9677-8f86b59a618b.png)

Instead of listing the app ID as description text, I now list the permissions that are allowed.

![Screenshot von 2020-02-22 09 19 43](https://user-images.githubusercontent.com/10796736/75089034-e7099480-5554-11ea-91d1-f46f6fc016f1.png)

I look forward to your feedback!
